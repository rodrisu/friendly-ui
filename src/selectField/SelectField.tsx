import React, { RefObject, useEffect, useState } from 'react'
import cc from 'classcat'

import { color } from '../_utils/branding'
import { A11yProps, CommonFieldsProps, pickA11yProps } from '../_utils/interfaces'
import { OnChangeParameters } from '../_utils/onChange'
import { ChevronIcon, ChevronIconDirections } from '../icon/chevronIcon'
import { StyledSelectField } from './SelectField.style'

export type SelectFieldItem = A11yProps &
  Readonly<{
    value: string | number
    label: string
  }>

export type SelectFieldProps = Partial<CommonFieldsProps> &
  A11yProps &
  Readonly<{
    options: Array<SelectFieldItem>
    defaultValue?: string
    onChange?: (obj: OnChangeParameters) => void
    onFocus?: (event: React.FocusEvent<HTMLSelectElement>) => void
    onBlur?: (event: React.FocusEvent<HTMLSelectElement>) => void
    focus?: boolean
    focusBorder?: boolean
    autoFocus?: boolean
    autoComplete?: string
  }>

export const SelectField = React.forwardRef(
  (props: SelectFieldProps, ref: RefObject<HTMLSelectElement>) => {
    const {
      name,
      className,
      defaultValue,
      options,
      onChange,
      onFocus = () => {},
      onBlur = () => {},
      disabled,
      required,
      focus,
      autoFocus,
      autoComplete,
      focusBorder = true,
    } = props
    const a11yAttrs = pickA11yProps<SelectFieldProps>(props)
    const baseClassName = 'kirk-selectField'
    const [hasFocus, setFocus] = useState(false)
    const [previousDefaultValue, setPreviousDefaultValue] = useState(defaultValue)
    const [currentValue, setCurrentValue] = useState(defaultValue)

    useEffect(() => {
      if (ref && !disabled && focus) {
        ref.current.focus()
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [disabled, focus])

    useEffect(() => {
      if (defaultValue !== previousDefaultValue) {
        setPreviousDefaultValue(defaultValue)
        setCurrentValue(defaultValue)
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [defaultValue])

    return (
      <StyledSelectField
        className={cc([
          baseClassName,
          className,
          hasFocus && focusBorder && `${baseClassName}--hasFocus`,
        ])}
      >
        <div className="kirk-selectField-background">
          <select
            name={name}
            onChange={event => {
              setCurrentValue(event.target.value)
              onChange({ name, value: event.target.value })
            }}
            onFocus={event => {
              setFocus(true)
              onFocus(event)
            }}
            onBlur={event => {
              setFocus(false)
              onBlur(event)
            }}
            defaultValue={defaultValue}
            value={currentValue}
            disabled={disabled}
            required={required}
            // eslint-disable-next-line jsx-a11y/no-autofocus
            autoFocus={autoFocus}
            autoComplete={autoComplete}
            ref={ref}
            {...a11yAttrs}
          >
            {options.map(({ value, label, ...optionA11yAttrs }: SelectFieldItem) => (
              <option key={`${value}${label}`} value={value} {...optionA11yAttrs}>
                {label}
              </option>
            ))}
          </select>
          <ChevronIcon
            iconColor={color.lightMidnightGreen}
            direction={ChevronIconDirections.DOWN}
          />
        </div>
      </StyledSelectField>
    )
  },
)
