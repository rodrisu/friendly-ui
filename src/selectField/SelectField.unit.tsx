import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'

import { SelectField as StyledSelectField } from './index'
import { SelectField } from './SelectField'

const options = [
  { value: '1', label: 'val 1', 'aria-label': 'value 1' },
  { value: '2', label: 'value 2' },
]

const defaultProps = {
  name: 'selectFieldName',
  defaultValue: options[0].value,
  options,
  onChange() {},
  onFocus() {},
  onblur() {},
}

const customProps = {
  ...defaultProps,
  id: 'selectFieldId',
  className: 'customClass',
  'aria-label': 'selectLabel',
  focus: true,
  autoFocus: true,
  disabled: false,
  required: true,
}

describe('SelectField', () => {
  const handleOnChange = jest.fn()
  const defaultWrapper = mount(<SelectField {...defaultProps} onChange={handleOnChange} />)

  it('should have the expected markup in the DOM', () => {
    const styledSelectField = renderer.create(<StyledSelectField {...defaultProps} />).toJSON()
    expect(styledSelectField).toMatchSnapshot()
  })

  it('should have proper initial value and number of options', () => {
    expect(defaultWrapper.find('select').props().defaultValue).toBe(options[0].value)
    expect(defaultWrapper.find('option')).toHaveLength(2)
  })

  it('should have appropriate alternate label on options', () => {
    expect(defaultWrapper.find('option').first().prop('aria-label')).toBe(options[0]['aria-label'])
    expect(defaultWrapper.find('option').last().prop('aria-label')).toBeUndefined()
  })

  it('should receive selected value after change', () => {
    const select = defaultWrapper.find('select')
    const selectDOMNode = select.getDOMNode()
    selectDOMNode.value = '2'
    select.simulate('change', { target: selectDOMNode })
    expect(handleOnChange).toHaveBeenCalledWith({ name: 'selectFieldName', value: '2' })
  })

  it('Should have custom props', () => {
    const wrapper = mount(<SelectField {...customProps} />)
    const selectField = wrapper.find('select')
    expect(wrapper.hasClass('customClass')).toBe(true)
    expect(wrapper.props().focus).toBe(true)
    expect(selectField.prop('id')).toEqual('selectFieldId')
    expect(selectField.prop('aria-label')).toEqual('selectLabel')
    expect(selectField.prop('autoFocus')).toBe(true)
    expect(selectField.prop('disabled')).toBe(false)
    expect(selectField.prop('required')).toBe(true)
  })

  describe('Focus event', () => {
    it('Simulates a focus event.', () => {
      expect(defaultWrapper.find('.kirk-selectField--hasFocus').exists()).toBe(false)
      defaultWrapper.find('select').simulate('focus')
      expect(defaultWrapper.find('.kirk-selectField--hasFocus').exists()).toBe(true)
    })

    it('Focus on mounting.', () => {
      defaultWrapper.setProps({ focus: true })
      expect(defaultWrapper.find('.kirk-selectField--hasFocus').exists()).toBe(true)
    })
  })

  describe('A11y', () => {
    it("Shouldn't be focused by default", () => {
      const selectField = defaultWrapper.find('select')
      expect(selectField.prop('autofocus')).toBe(undefined)
    })
  })
})
