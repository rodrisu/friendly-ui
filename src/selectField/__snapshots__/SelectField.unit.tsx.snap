// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField should have the expected markup in the DOM 1`] = `
.c1 {
  fill: #708C91;
}

.c1.kirk-icon-wrapper {
  display: inline-block;
  position: relative;
}

.c1.kirk-icon-shadowed {
  border-radius: 50%;
  box-shadow: 0 0 2px rgba(0,0,0,.3);
}

.c1 .kirk-icon-badge {
  position: absolute;
  bottom: 12px;
  left: 12px;
}

.c0 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  padding-left: 24px !important;
  padding-right: 24px !important;
  margin-left: 0;
  margin-right: 0;
  padding: 8px 0;
}

.c0 .kirk-selectField-background {
  position: relative;
  color: #054752;
  background-color: #EDEDED;
  border-radius: 16px;
  border: solid 1px #EDEDED;
  box-shadow: none;
  padding: 0 0 0 16px;
}

.c0 select {
  display: block;
  position: relative;
  width: 100%;
  height: 52px;
  margin: 0;
  padding: 0 calc(8px + 24px) 0 0;
  font-size: 16px;
  line-height: 20px;
  color: #054752;
  box-shadow: none;
  background: transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  outline: 0;
  box-shadow: none;
  border: none;
  border-radius: 16px;
  cursor: pointer;
  z-index: 1;
}

.c0 select::-ms-expand {
  display: none;
}

.c0 select:focus {
  outline: none;
}

.c0 .kirk-icon {
  position: absolute;
  top: 0;
  bottom: 0;
  margin: auto;
  right: 8px;
  background: #EDEDED;
  z-index: 0;
}

.c0 .kirk-selectField--hasFocus {
  border: 3px solid #00AFF5;
}

.c0 .kirk-selectField--hasFocus select {
  height: calc(52px - 1px - 3px);
}

<div
  className="c0 kirk-selectField"
>
  <div
    className="kirk-selectField-background"
  >
    <select
      defaultValue="1"
      name="selectFieldName"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="1"
    >
      <option
        aria-label="value 1"
        value="1"
      >
        val 1
      </option>
      <option
        value="2"
      >
        value 2
      </option>
    </select>
    <svg
      aria-hidden={true}
      className="kirk-icon c1"
      height={24}
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      
      <polyline
        fill="none"
        points="9 18 15 12 9 6"
        stroke="#708C91"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeMiterlimit="10"
        strokeWidth="2"
        transform="rotate(90 12 12)"
      />
    </svg>
  </div>
</div>
`;
